### ISSUE TITLE
CCE rollout is duplicated across 5 envs — drift has already caused two defects

### ISSUE THREAD

**Issue by manchado53** [manchado53/counterfactual-reasoning#12]
The counterfactual rollout that produces the CCE score is written **five separate times**, once
per environment. The skeleton is identical; only the env step and the action rule differ. Drift
between the copies has already happened and has already cost us (see the linked issue below).

## The five copies

| env | file | continuation policy | opponent in the loop? | net input |
|---|---|---|---|---|
| JaxNav | `agents/jax_nav/consequence_dqn.py` | **softmax(q / 0.5)** | no | `obs` |
| FrozenLake | `agents/frozen_lake/consequence_dqn.py` | argmax | no | state *is* the input |
| SMAX | `agents/smax/consequence_dqn.py` | argmax (masked) | no | `world_state` |
| Connect Four | `agents/connect_four/consequence_dqn.py` | argmax (masked) | random | `obs` + legal mask |
| chess | `agents/chess/consequence_dqn.py` | argmax (masked) | `opp_step` | `obs` + legal mask |

## What is genuinely identical in all five

- reward masking after termination — `jnp.where(done, 0.0, r)`
- discount freeze on termination — `jnp.where(done, disc, disc * gamma)`
- done propagation — `done | new_done`
- scan length — `horizon - 1`, with the first action applied before the scan and undiscounted
- carry shape — `(state, [obs], key, cum, disc, done)`, returning the cumulative entry
- the triple `vmap` (rollouts → actions → states) and the `jax.jit` around it

That is ~8 lines duplicated 5x, and it is the core scoring path of every experiment in the repo.

## Why this is not cosmetic

Two defects already trace directly to the duplication:

- `cf_rollout_temperature` exists **only** in JaxNav. FrozenLake-deterministic needs it for the
  same reason and never got it — filed separately.
- #3 — aggregation silently falls back to `max` on FrozenLake.

A fix or an audit currently has to be applied five times and it is easy to miss one.

## Proposed shape (NOT before the ICLR deadline)

A shared skeleton in `agents/shared/` taking an env-supplied step function:

```
run_cf_rollout(step_fn, policy_fn, first_action, state, key, horizon, gamma, ...)
```

Connect Four and chess step an *opponent* inside the scan, so `step_fn` has to be allowed to
advance more than one ply per tick — that is the main interface constraint.

## Deliberately deferred

This is the scoring core of every result in the paper. A silent refactor bug here corrupts
everything without failing loudly, and the abstract deadline is 2026-09-18. Filing it so it is
not forgotten, not to do it now.

When it is done, it needs a golden-value regression test per env: score a fixed checkpoint on
fixed states before and after, and require bit-identical output.
